RavenDB-14541:Added RavenQuery.Include

aharonlevy opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-14541/Add-RavenQuery.Include

### Additional description

Added RavenQuery.Include

### Type of change
- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

Commit:
RavenDB-14541: added RavenQuery.Include/IncludeCounters/IncludeAllCounters

ayende commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3109,6 +3115,18 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
 
+            if (js.StartsWith("include("))
+            {
+                Regex isDiscard = new Regex("_+", RegexOptions.IgnoreCase);
```
This creates a new Regex each time. Create a static instance with `Compiled`, instead.

For that matter, why are you using a regex instead of a `Contains` or `IndexOf` ? 

What is this meant to catch?

aharonlevy replied:
Done.
(Edited the Regex to a static instance with `Compiled`.)

We try catching what the var name given by the user when calling `RavenQuery.Include`

For example:
`var query = from u in session.Query<Order>()
                        let **includes** = RavenQuery.Include<Order>(u.Employee)
                        select new QueryResult { Comapny = u.Company };`

We found that it'll return an array of null values when we use `query.toList()`.
After consulting with Pawel, we decided to force the user to use '_' or '__', '___' etc... and throw when a user tries using something else.
So it meant to make sure the user made a valid query. I think `contains` and `IndexOf` are not suitable for it

ayende commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3211,6 +3229,13 @@ private string TranslateSelectBodyToJs(Expression expression)
 
+            if (expression is MemberExpression memberExpression)
+            {
+                var name = GetSelectPath(memberExpression.Member);
```
??

aharonlevy replied:
I have double-checked it. The code was unnecessary, so I removed it.

ayende commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,126 @@
+        catch (Exception e)
+        {
+            Assert.NotNull(e);
+            Assert.Contains("You can't use the include that way try: let _= RavenQuery.Include<T>()", e.InnerException.Message);
```
Why is this an error?

aharonlevy replied:
The error is using a wrong name in the let when using 'RavenQuery.Include'.

The expected query is supposed to be as follows:
`var query = from u in session.Query<Order>()
 let _ = RavenQuery.Include<Order>(u.Employee) 
select new QueryResult { Comapny = u.Company };`

garayx replied:
we don't want the user to assign the included value to a variable, since it is included to session. So we allow only to use `let _+`

ayende replied:
Why are we bothering with that at all?
If they try to assign the variable, they will get no results, right?
An array of `null` ? 

We can make this return the actual document ids, by the way, and that would be far cleaner.

ayende commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,126 @@
+                {
+
+                    var query3 = from u in session.Query<Order>()
+                        let includes = RavenQuery.Include<Order>(u.Employee)
```
What does this give us that `Include(x=>x.Employee)` doesn't?

ayende commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,126 @@
+            using (var session = store.OpenSession())
+            {
+                var query3 = from a in session.Query<Address>()
+                             let _ = RavenQuery.Include<State>(a.CountryState.Split('#',StringSplitOptions.None)[0])
```
Can we try making this work in `Include(x=>x..CountryState.Split('#',StringSplitOptions.None)[0])` instead?

aharonlevy replied:
It's possible. Do we want to keep the current implementation that enables:
`Include<Address>(x.CountryState.Split('#', StringSplitOptions.None)[0])`

aharonlevy replied:
done

garayx commented on src/Raven.Client/Documents/Queries/RavenQuery.cs:
```diff
@@ -18,6 +20,26 @@ public static IEnumerable<T> Load<T>(IEnumerable<string> ids)
             throw new NotSupportedException("This method is here for strongly type support of server side call during Linq queries and should never be directly called");
         }
 
+        public static  bool? Include(string path)
```
formatting

aharonlevy replied:
done

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3120,10 +3120,23 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+            if (js.StartsWith("includeCounters("))
+            {
+                if (!IsDiscard.IsMatch(name))
+                    throw new NotSupportedException("You can't use the include that way try: let _= RavenQuery.IncludeTimeSeries(<T> documentInstance, string timeSeriesName)");
```
should be include counters

aharonlevy replied:
done

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3109,13 +3116,44 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+            if (js.StartsWith("includetimeSeries("))
+            {
+                if (!IsDiscard.IsMatch(name))
+                    throw new NotSupportedException("You can't use the include that way try: let _= RavenQuery.IncludeTimeSeries(<T> documentInstance, string timeSeriesName)");
```
```suggestion
                    throw new NotSupportedException("You can't use the include that way try: let _= RavenQuery.IncludeTimeSeries(T documentInstance, string timeSeriesName)");
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,149 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                            writer.Write(")");
+                            return;
+                        }
+                        throw new NotSupportedException("You can't use the include that way try: let _= RavenQuery.Include<T>(\"path\")");
```
```suggestion
                        throw new NotSupportedException("You can't use the include that way try: let _ = RavenQuery.Include<T>(\"path\")");
```

also suggest the other overloads

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,149 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+
+                        writer.Write("include(");
+                        context.Visitor.Visit(targetExpression);
+                        writer.Write(")");
```
do we want to `return;` here ?

aharonlevy replied:
We do. I thought it'd be extra because it's the end of the method.
added `return` here

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,149 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                        writer.Write("include(");
+                        context.Visitor.Visit(targetExpression);
+                        writer.Write(")");
+                    }
```
what happens if we didn't enter any of the if clauses above ?

aharonlevy replied:
Then we should throw, I think.

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,149 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+            {
+                var methodCallExpression = context.Node as MethodCallExpression;
+
+                if (methodCallExpression?.Method.Name != nameof(RavenQuery.IncludeTimeSeries))
```
should be IncludeCounters

aharonlevy replied:
done

garayx commented on src/Raven.Server/Documents/Queries/Parser/QueryParser.cs:
```diff
@@ -1319,7 +1319,23 @@ private DeclaredFunction DeclaredFunction()
                 // validate this function
                 try
                 {
+                    var p1 = Scanner.Position;
+                    var x = Scanner.Position - funcBodyStart - 2;
```
lets have better naming, something like:
```suggestion
                    var pos = Scanner.Position;
                    var includePos = Scanner.Position - funcBodyStart - 2;
```

aharonlevy replied:
done

garayx commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,254 @@
+﻿
```
formatting

aharonlevy replied:
done

garayx commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,254 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Confluent.Kafka;
```
are all those necessary?

aharonlevy replied:
No, deleted the extras.

garayx commented on test/Tryouts/Program.cs:
```diff
@@ -29,9 +38,9 @@ public static async Task Main(string[] args)
+                    using (var test = new QueryOnCounters(testOutputHelper))
                     {
-                        await test.ContinueFromThePointIStoppedConcurrentSubscription(1, 20);
+                         /*test.SessionQuerySelectAdressFromIncludeDoc_UsingRavenQueryCounter();*/
```
debug code

aharonlevy replied:
done

garayx commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,247 @@
+                using (var session = store.OpenSession())
+                {
+                    var query3 = from a in session.Query<Address>()
+                        let _ = RavenQuery.Include<Address>(a.StateId)
```
this is the path case ? I think the correct syntax would be this?
`let _ = RavenQuery.Include<Address>(a => a.StateId)`

also we should forbid such syntax, and allow either string or lambda

aharonlevy replied:
This is a test that shows we don't allow such syntax; it'll throw and then have an error message with the following:
`You can't use the include that way try: let _ = RavenQuery.Include<T>(Expression<Func<T, string>> path)`

garayx replied:
its not really testing that it throws, you should wrap the code that should throw in Assert.Throws, something like this:

https://github.com/ravendb/ravendb/blob/de8a154c3ebf19f2ad1faf6e1e97b2024cac2eb3/test/EmbeddedTests/RuntimeFrameworkVersionMatcherTests.cs#L86

garayx commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,247 @@
+            using (var session = store.OpenSession())
+            {
+                var query3 = from a in session.Query<Address>()
+                    let _ = RavenQuery.Include<State>("a.StateId")
```
should be `let _ = RavenQuery.Include<State>("StateId")`

aharonlevy replied:
done

garayx commented on test/SlowTests/Issues/RavenDB-14541.cs:
```diff
@@ -0,0 +1,247 @@
+            using (var session = store.OpenSession())
+            {
+                var query3 = from a in session.Query<Address>()
+                    let _ = RavenQuery.Include("a.StateId")
```
should be `let _ = RavenQuery.Include("StateId")` ?

aharonlevy replied:
done

garayx commented on test/SlowTests/Issues/RavenDB-14541_includeTimeSeries.cs:
```diff
@@ -0,0 +1,499 @@
+                        let _ = RavenQuery.IncludeTimeSeries(a, "CurViews")
+                        let __ = RavenQuery.IncludeTimeSeries(a, "CurViews2")
+                        select new { Name = a.UserName };
+                    // TODO: add test: 
```
debug code

aharonlevy replied:
done

garayx commented on test/SlowTests/Issues/RavenDB-14541_includeTimeSeries.cs:
```diff
@@ -0,0 +1,499 @@
+                    var query = from a in session.Query<User>()
+                                let _ = RavenQuery.IncludeAllCounters(a)
+                                let __ = RavenQuery.IncludeTimeSeries(a, "CurViews")
+                                let a_ = RavenQuery.Include<State>("a.StateId")
```
why `let a_` is allowed? we should allow only discard variables

aharonlevy replied:
I had a wrong regex, fixed the regex, and the test

garayx replied:
please make a test that will throws on `let a_ `

garayx commented on test/Tryouts/Program.cs:
```diff
@@ -29,9 +28,9 @@ public static async Task Main(string[] args)
+                    using (var test = new RavenDB_14541_IncludeTimeSeries(testOutputHelper))
                     {
-                        await test.ContinueFromThePointIStoppedConcurrentSubscription(1, 20);
+                        test.SessionQueryincludeTimeSeriesUsingRavenQuery_useingTwoLets();
```
debug code

ppekrol commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,185 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                        var replaceExprassion = Expression.Parameter(targetExpression.Expression.Type, _alias);
+                        targetExpression = targetExpression.Update(replaceExprassion);
+
+                        writer.Write("include(");
```
this is legacy syntax as clearly is stated in `ScriptRunner`. Please use `includes.document`

aharonlevy replied:
done

ppekrol commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,185 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                using (writer.Operation(methodCallExpression))
+                {
+                    int argNum = methodCallExpression.Arguments.Count;
+                    writer.Write("includetimeseries(");
```
`includes.timeseries(document, 'name')`

aharonlevy replied:
done

ppekrol commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,185 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                using (writer.Operation(methodCallExpression))
+                {
+                    int argNum = methodCallExpression.Arguments.Count;
+                    writer.Write("includecounters(");
```
`includes.counters(document, 'name')`

aharonlevy replied:
done

ppekrol commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,185 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                using (writer.Operation(methodCallExpression))
+                {
+                    int argNum = methodCallExpression.Arguments.Count;
+                    writer.Write("includeallcounters(");
```
`includes.counters(document)`

aharonlevy replied:
Deleted IncludeAllCounterSupport class because we made the same call in includeCountersSupport and it'll already be made from there.
done

ppekrol commented on src/Raven.Server/Documents/Queries/Parser/QueryParser.cs:
```diff
@@ -75,7 +75,7 @@ public bool TryParse(out Query query, out string message, QueryType queryType =
             while (Scanner.TryScan("DECLARE"))
             {
-                var func = DeclaredFunction();
+                var func = DeclaredFunction(query);
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/Parser/QueryParser.cs:
```diff
@@ -1271,11 +1271,11 @@ private List<QueryExpression> IncludeClause()
                     ThrowParseException("Unable to understand include clause expression");
                 }
-            } while (Scanner.TryScan(","));
+            } while (Scanner.TryScan(",") || Scanner.TryScan(";"));
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/Parser/QueryParser.cs:
```diff
@@ -1271,11 +1271,11 @@ private List<QueryExpression> IncludeClause()
         }
 
-        private DeclaredFunction DeclaredFunction()
+        private DeclaredFunction DeclaredFunction(Query query)
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/Parser/QueryParser.cs:
```diff
@@ -1319,7 +1319,22 @@ private DeclaredFunction DeclaredFunction()
+                    var includePos = Scanner.Position - funcBodyStart - 2;
+
                     var program = ValidateScript(functionText);
+                    if (functionText.Contains("include"))
```
this is not te way to do it, revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/Parser/QueryParser.cs:
```diff
@@ -1359,6 +1374,11 @@ private DeclaredFunction DeclaredFunction()
             };
         }
 
+        private bool isInclude(QueryParser p)
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -980,7 +980,16 @@ private void AddToTimeSeriesIncludes(TimeSeriesIncludesField timeSeriesIncludes,
                         if (!(expression.Arguments[0] is MethodExpression methodExpression))
-                            throw new InvalidQueryException($"Expected to get include '{nameof(MethodType.TimeSeries)}' clause expression, but got: '{expression}'.", Query.QueryText, parameters);
+                        {
+                            if (expression.Arguments[1] is ValueExpression vt)
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMethod.cs:
```diff
@@ -97,6 +97,12 @@ public static MethodType GetMethodType(string methodName, bool throwIfNoMatch =
             if (string.Equals(methodName, "counters", StringComparison.OrdinalIgnoreCase))
                 return MethodType.Counters;
 
+            if (string.Equals(methodName, "includecounters", StringComparison.OrdinalIgnoreCase))
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMethod.cs:
```diff
@@ -97,6 +97,12 @@ public static MethodType GetMethodType(string methodName, bool throwIfNoMatch =
+            if (string.Equals(methodName, "includecounters", StringComparison.OrdinalIgnoreCase))
+                return MethodType.Counters;
+
+            if (string.Equals(methodName, "includeallcounters", StringComparison.OrdinalIgnoreCase))
```
revert

aharonlevy replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMethod.cs:
```diff
@@ -106,6 +112,9 @@ public static MethodType GetMethodType(string methodName, bool throwIfNoMatch =
             if (string.Equals(methodName, "timeseries", StringComparison.OrdinalIgnoreCase))
                 return MethodType.TimeSeries;
 
+            if (string.Equals(methodName, "includetimeseries", StringComparison.OrdinalIgnoreCase))
```
revert

aharonlevy replied:
done

ppekrol reviewed (changes requested):
The approach here is incorrect.

Following query should be generated by the Client API:
```
declare function output(a) {
	includes.timeseries(a,"CurViews");
	return { Name : a.UserName };
}
from 'Users' as a select output(a)
```

All necessary changes have to be done in `ScriptRunner` around
```
                // includes
                var includeDocumentFunc = new ClrFunctionInstance(ScriptEngine, "include", IncludeDoc);
                ObjectInstance includesObject = new ObjectInstance(ScriptEngine);
                includesObject.FastAddProperty("document", includeDocumentFunc, false, false, false);
                includesObject.FastAddProperty("cmpxchg", new ClrFunctionInstance(ScriptEngine, "cmpxchg", IncludeCompareExchangeValue), false, false, false);
                includesObject.FastAddProperty("revisions", new ClrFunctionInstance(ScriptEngine, "revisions", IncludeRevisions), false, false, false);
                ScriptEngine.SetValue("includes", includesObject);
```

those line of code.

There are few Qs to answer:
1. RavenQuery.Include<Order>(u.Employee) should probably return `string[]` with IDs of the included documents, then we do not have to worry about invalid usage of the API
2. What needs to be returned for other methods? Please figure it out.

Commit:
RavenDB-14541:fixed the bug that failed the tests

Commit:
RavenDB-14541: Removed whitespaces

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,10 +3136,39 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+            }
+        }
+
+        void HandleInclude(string name, string js, string message)
```
```suggestion
        void HandleRavenQueryInclude(string name, string js, string message)
```

aharonlevy replied:
done

aharonlevy replied:
Changed the function so it'll handle regulars variables as well so changed it to `HandleFunctionBody`

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,10 +3136,39 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+            AppendIncludeFunctionBody(js);
+        }
+
+        private void AppendIncludeFunctionBody(string js)
```
this method is only used in `HandleInclude`

aharonlevy replied:
moved into `HandleInclude`

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,221 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
             }
         }
 
+        public class IncludeSupport : JavascriptConversionExtension
```
```suggestion
        public class IncludeDocumentSupport : JavascriptConversionExtension
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,10 +3136,39 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+                    HandleInclude(name, s, "let _ = RavenQuery.IncludeTimeSeries(T documentInstance, string timeSeriesName)");
+                    break;
+                case string s when s.StartsWith("includes.counters("):
+                    HandleInclude(name, s, "let _ = RavenQuery.includeCounters(T documentInstance, string CounterName)");
```
`RavenQuery.includeCounters` is not correct syntax

aharonlevy replied:
Changed the switch to the single if statement that we talked about.

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,10 +3136,39 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+                    HandleInclude(name, s, "let _ = RavenQuery.includeCounters(T documentInstance, string CounterName)");
+                    break;
+                default:
+                    _declareBuilder ??= new StringBuilder();
```
this is almost same code as `AppendIncludeFunctionBody` can we reuse it?

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,221 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+            }
+        }
+
+        public class IncludeCounterSupport : JavascriptConversionExtension
```
the code here is almost the same as other Counter JavascriptConversionExtension can we create a single JavascriptConversionExtension for all counter variants?

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,221 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                {
+                    int argNum = methodCallExpression.Arguments.Count;
+                    writer.Write("includes.counters(");
+                    for (var i = 0; i < argNum; i++)
```
it better to use boolean to write the comma:
something like this:
```

                    writer.Write("includes.counters(");
                    var first = true;
                    for (var i = 0; i < methodCallExpression.Arguments.Count; i++)
                    {
                        context.Visitor.Visit(methodCallExpression.Arguments[i]);
                        if (first == false){
                            writer.Write(",");
                         }
                         first = false;
                    }
```

aharonlevy replied:
done

garayx commented on src/Raven.Server/Documents/Queries/Dynamic/CollectionQueryRunner.cs:
```diff
@@ -173,8 +176,8 @@ private async ValueTask ExecuteCollectionQueryAsync(QueryResultServerSide<Docume
 
-                IncludeCountersCommand includeCountersCommand = null;
-                IncludeTimeSeriesCommand includeTimeSeriesCommand = null;
+                    //includeCountersCommand = null;
```
debug code

aharonlevy replied:
done

garayx commented on src/Raven.Server/Documents/Queries/GraphQueryRunner.cs:
```diff
@@ -145,8 +147,9 @@ public async Task WriteDetailedQueryResult(IndexQueryServerSide indexQuery, Quer
                         idc = new IncludeDocumentsCommand(Database.DocumentsStorage, queryContext.Documents, query.Metadata.Includes, fieldsToFetch.IsProjection);
                         icevc = IncludeCompareExchangeValuesCommand.ExternalScope(queryContext, query.Metadata.CompareExchangeValueIncludes);
                         irc = new IncludeRevisionsCommand(database: Database, context: queryContext.Documents, query.Metadata.RevisionIncludes);
+                        itsc = new IncludeTimeSeriesCommand(context: queryContext.Documents, query.Metadata.TimeSeriesIncludes?.TimeSeries ?? null);
+                        icc = new IncludeCountersCommand(database: Database, context: queryContext.Documents, query.Metadata.CounterIncludes?.Counters ?? null);
```
I think we can leave those as nulls

aharonlevy replied:
done

garayx commented on src/Raven.Server/Documents/Includes/IncludeCountersCommand.cs:
```diff
@@ -37,7 +37,24 @@ public IncludeCountersCommand(DocumentDatabase database, DocumentsOperationConte
             : this(database, context)
         {
-            _countersBySourcePath = countersBySourcePath.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToArray());
+            _countersBySourcePath = countersBySourcePath?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToArray())?? new Dictionary<string, string[]>();
```
is there a place that countersBySourcePath can be null?

aharonlevy replied:
fixed as we talked

Commit:
RavenDB-14541:removed whitespace

garayx commented on src/Raven.Server/Documents/Queries/Dynamic/CollectionQueryRunner.cs:
```diff
@@ -137,7 +138,9 @@ private async ValueTask ExecuteCollectionQueryAsync(QueryResultServerSide<Docume
                 var includeDocumentsCommand  = new IncludeDocumentsCommand(Database.DocumentsStorage, context.Documents, query.Metadata.Includes, fieldsToFetch.IsProjection);
                 var includeRevisionsCommand  = new IncludeRevisionsCommand(Database, context.Documents, query.Metadata.RevisionIncludes);
-                
+                var includeTimeSeriesCommand = new IncludeTimeSeriesCommand(context.Documents, new Dictionary<string, HashSet<AbstractTimeSeriesRange>>());
+                var includeCountersCommand = new IncludeCountersCommand(Database, context.Documents, new Dictionary<string, HashSet<string>>());
+
```
we missing the comparer here

aharonlevy replied:
done

Commit:
RavenDB-14541: merged all IncludeCountersSupport and IncludeTimeSeriesSupport to one class
changed where Reasult is initialized in IncludeCountersCommand and IncludeTimeSeriesCommand

Commit:
RavenDB-14541: converted to '?:' expression

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,11 +3136,27 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+                {
+                    throw new InvalidOperationException("Wrong syntax, 'RavenQuery.Include()' works only with discard variable.");
+                }
+                HanadleFunctionBOdy(name, js, isInclude: true);
```
```suggestion
                HandleFunctionBody(name, js, isInclude: true);
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,11 +3136,27 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
-                .Append(" = ").Append(js).Append(';')
-                .Append(Environment.NewLine);
+            _declareBuilder.Append('\t');
+            if (!isInclude)
```
```suggestion
            if (isInclude == false)
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,163 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                    {
+                        var arg = methodCallExpression.Arguments[0];
+
+                        if (arg is ConstantExpression { Value: string target1})
```
```suggestion
                        if (arg is ConstantExpression { Value: string target1 })
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,163 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                            writer.Write(")");
+                            return;
+                        }
+                        throw new NotSupportedException("You can't use the include that way try: let _ = RavenQuery.Include<T>(Expression<Func<T, string>> path)");
```
lets add the current expression to the exception

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,163 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                        writer.Write(")");
+                        return;
+                    }
+                    throw new NotSupportedException("You can't use the include that way try: let _ = RavenQuery.Include<T>(Expression<Func<T, string>> path)");
```
lets add the current expression to the exception

aharonlevy replied:
done

Commit:
RavenDB-14541: made more readable, removed debug text

Commit:
RavenDB-14541: removed extra new line

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,11 +3136,27 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+            {
+                if (IsDiscard.IsMatch(name) == false)
+                {
+                    throw new InvalidOperationException($"Wrong syntax, 'RavenQuery.Include()' works only with discard variable. current variable is called {name}");
```
```suggestion
                    throw new InvalidOperationException($"Wrong syntax, 'RavenQuery.Include()' works only with discard variable. Current variable is called '{name}'.");
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -3129,11 +3136,27 @@ private void AppendLineToOutputFunction(string name, string js, StringBuilder wr
+            }
+        }
+
+        void HandleFunctionBOdy(string name, string js, bool isInclude = false)
```
```suggestion
        void HandleFunctionBody(string name, string js, bool isInclude = false)
```

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,129 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+
+                var hasTimeSeries = methodName == nameof(RavenQuery.IncludeTimeSeries);
+
+                    if (hasCounters == false && hasTimeSeries == false)
+                    return;
```
code format

aharonlevy replied:
done

garayx commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -1008,6 +1008,129 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                    if (hasCounters == false && hasTimeSeries == false)
+                    return;
+
+                var expectedArgsCount = methodName == nameof(RavenQuery.IncludeAllCounters) ? 1 : 2;
```
make a variable from `methodName == nameof(RavenQuery.IncludeAllCounters)` (like `hasCounters`) and use it here and in `hasCounters`

aharonlevy replied:
done

garayx commented on src/Raven.Server/Documents/Patch/ScriptRunner.cs:
```diff
@@ -895,6 +899,100 @@ private JsValue IncludeCompareExchangeValue(JsValue self, JsValue[] args)
+            private JsValue IncludeTimeSeriesValue(JsValue self, JsValue[] args)
+            {
+                if (args.Length != 2)
+                    throw new InvalidOperationException($"includes.timeseries(doc, name) must be called with two arguments. now called with {args.Length} arguments");
```
```suggestion
                    throw new InvalidOperationException($"includes.timeseries(doc, name) must be called with two arguments. Now called with '{args.Length}' arguments.");
```

aharonlevy replied:
done

garayx commented on src/Raven.Server/Documents/Queries/GraphQueryRunner.cs:
```diff
@@ -157,7 +157,9 @@ public async Task WriteDetailedQueryResult(IndexQueryServerSide indexQuery, Quer
                             icevc,
-                            irc);
+                            irc,
+                            includeTimeSeriesCommand:null,
+                            includeCountersCommand:null);
```
```suggestion
                            includeTimeSeriesCommand: null,
                            includeCountersCommand: null);
```

aharonlevy replied:
done

Commit:
RavenDB-14541: fixed as review asked

Commit:
RavenDB-14541:Formatting

arekpalinski commented:
Pushed the changed to https://github.com/ravendb/ravendb/tree/RavenDB-14541 branch. Closing the PR.
